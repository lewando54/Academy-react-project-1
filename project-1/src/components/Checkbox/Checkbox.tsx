import React from 'react';
import {useState, useEffect} from 'react';

interface CheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
    children: string,
    id: string,
    onChecked: Function
}
// Checkbox powinien mieÄ‡ stan checked (kontrolowany komponent)
function Checkbox({id, children, onChecked, ...props}: CheckboxProps){
    const [checked, setChecked] = useState(false);

    const onChangeHandler = () => {
        setChecked(prev => !prev);
    }

    useEffect(() => {
        onChecked(checked);
    }, [checked]);

    return (
        <div className='w-full flex gap-1 justify-start'>
            <input id={id} type="checkbox" onChange={onChangeHandler} {...props}/>
            <label htmlFor={id}>{children}</label>
        </div>
    )
}

export default Checkbox;