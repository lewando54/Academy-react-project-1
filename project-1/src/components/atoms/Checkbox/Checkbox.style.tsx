import styled from 'styled-components'

const StyledCheckbox = styled.input`
    -webkit-appearance: none;
    appearance: none;
    width: 1.6em;
    height: 1.6em;
    border-radius: 0.15em;
    margin-right: 0.5em;
    border: 0.15em solid #e5e7eb;
    outline: none;
    cursor: pointer;
    transition: all 0.2s ease-in-out;

    &:before {
        content: "";
        font-size: 1.5em;
        color: #fff;
        position: absolute;
        right: 1px;
        top: -5px;
        transform: translateX(-3px) translateY(4px) rotateZ(180deg) rotateY(180deg);
        clip-path: polygon(100% 0%, 100% 0%, 100% 100%, 100% 100%);
        transition: clip-path 0.3s ease-in-out;
    }

    &:hover{
        border: 0.15em solid ${props => (props.color != null) ? props.color : 'var(--primary)'};
    }

    &:checked {
        border: 0.15em solid ${props => (props.color != null) ? props.color : 'var(--primary)'};
        background-color: ${props => (props.color != null) ? props.color : 'var(--primary)'};
        position: relative;
        display: flex;
        justify-content: center;
        align-items: center;
    }

    &:checked:before {
        content: "L";
        font-size: 1rem;
        font-weight: bold;
        color: #fff;
        position: absolute;
        right: 1px;
        top: -5px;
        clip-path: polygon(0% 0%, 100% 0%, 100% 100%, 0% 100%);
        transform: translateX(-3px) translateY(4px) rotateZ(45deg) rotateY(180deg);
    }
`

export default StyledCheckbox
