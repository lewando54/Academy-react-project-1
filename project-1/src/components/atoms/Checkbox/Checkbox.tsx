import React, { forwardRef } from 'react'
import StyledCheckbox from './Checkbox.style'

export interface CheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
  id: string
  color?: string
  testId: string
}

function Checkbox ({ id, color, testId }: CheckboxProps, ref: React.Ref<HTMLInputElement>): React.ReactElement {
  return (
        <StyledCheckbox data-testid={testId} ref={ref} id={id} type={'checkbox'} color={color}/>
  )
}

export default forwardRef(Checkbox)
