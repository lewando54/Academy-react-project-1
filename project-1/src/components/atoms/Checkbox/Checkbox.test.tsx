import React from 'react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import * as renderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react'
import Checkbox from './Checkbox'

describe('Checkbox', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Checkbox testId="checkbox" id="test" />)
    expect(tree).toMatchSnapshot()
  })

  it('should be checked if clicked when is unchecked', () => {
    const { getByTestId } = render(<Checkbox testId="checkbox" id="test" />)
    fireEvent.click(getByTestId('checkbox'))
    expect(getByTestId('checkbox')).toBeChecked()
  })

  it('should be unchecked if clicked twice', () => {
    const { getByTestId } = render(<Checkbox testId="checkbox" id="test" />)
    fireEvent.click(getByTestId('checkbox'))
    expect(getByTestId('checkbox')).toBeChecked()
    fireEvent.click(getByTestId('checkbox'))
    expect(getByTestId('checkbox')).not.toBeChecked()
  })

  it('should have correct color on hover', () => {
    const { getByTestId } = render(<Checkbox id='checkbox' color='red' testId='checkbox'/>)
    expect(getByTestId('checkbox')).toHaveStyleRule('border: 0.15em solid red')
  })

  it('should have correct color when checked', () => {
    const { getByTestId } = render(<Checkbox id='checkbox' color='red' testId='checkbox'/>)
    fireEvent.click(getByTestId('checkbox'))
    expect(getByTestId('checkbox')).toHaveStyle({ backgroundColor: 'red' })
  })
})
