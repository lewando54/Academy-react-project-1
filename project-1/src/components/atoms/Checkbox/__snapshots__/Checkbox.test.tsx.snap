// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render correctly 1`] = `
.c0 {
  --primary: rgb(236 72 153);
  --secondary: rgb(229 231 235);
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.6em;
  height: 1.6em;
  border-radius: 0.15em;
  margin-right: 0.5em;
  border: 0.15em solid #e5e7eb;
  outline: none;
  cursor: pointer;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.c0:before {
  content: "";
  font-size: 1.5em;
  color: #fff;
  position: absolute;
  right: 1px;
  top: -5px;
  -webkit-transform: translateX(-3px) translateY(4px) rotateZ(180deg) rotateY(180deg);
  -ms-transform: translateX(-3px) translateY(4px) rotateZ(180deg) rotateY(180deg);
  transform: translateX(-3px) translateY(4px) rotateZ(180deg) rotateY(180deg);
  -webkit-clip-path: polygon(100% 0%,100% 0%,100% 100%,100% 100%);
  clip-path: polygon(100% 0%,100% 0%,100% 100%,100% 100%);
  -webkit-transition: clip-path 0.3s ease-in-out;
  transition: clip-path 0.3s ease-in-out;
}

.c0:hover {
  border: 0.15em solid var(--primary);
}

.c0:checked {
  border: 0.15em solid var(--primary);
  background-color: var(--primary);
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:checked:before {
  content: "L";
  font-size: 1rem;
  font-weight: bold;
  color: #fff;
  position: absolute;
  right: 1px;
  top: -5px;
  -webkit-clip-path: polygon(0% 0%,100% 0%,100% 100%,0% 100%);
  clip-path: polygon(0% 0%,100% 0%,100% 100%,0% 100%);
  -webkit-transform: translateX(-3px) translateY(4px) rotateZ(45deg) rotateY(180deg);
  -ms-transform: translateX(-3px) translateY(4px) rotateZ(45deg) rotateY(180deg);
  transform: translateX(-3px) translateY(4px) rotateZ(45deg) rotateY(180deg);
}

<input
  className="c0"
  data-testid="checkbox"
  id="test"
  type="checkbox"
/>
`;
